Makefile

```diff
@@ -16,75 +15,141 @@ BRIGADE_GITHUB_GW_EXTERNAL_PORT := 7744
-CX_ARCHS = amd64
-
-# Build native binaries
-.PHONY: build
-build: $(BINS)
-
-.PHONY: $(BINS)
-$(BINS): vendor
-	go build -ldflags '$(LDFLAGS)' -o bin/$@ ./$@/cmd/$@
-
-# Cross-compile for Docker+Linux
-build-docker-bins: $(addsuffix -docker-bin,$(BINS))
-
-%-docker-bin: vendor
-	GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o ./$*/rootfs/$* ./$*/cmd/$*
-
-# To use docker-build, you need to have Docker installed and configured. You should also set
-# DOCKER_REGISTRY to your own personal registry if you are not pushing to the official upstream.
-.PHONY: docker-build
-docker-build: build-docker-bins
-docker-build: $(addsuffix -image,$(IMAGES))
-
-%-image:
-	docker build $(DOCKER_BUILD_FLAGS) -t $(DOCKER_REGISTRY)/$*:$(IMAGE_TAG) $*
-
-# You must be logged into DOCKER_REGISTRY before you can push.
-.PHONY: docker-push
-docker-push: $(addsuffix -push,$(IMAGES))
-
-%-push:
-	docker push $(DOCKER_REGISTRY)/$*:$(IMAGE_TAG)
-
-# Cross-compile binaries for our CX targets.
-# Mainly, this is for brig-cross-compile
-%-cross-compile: vendor
-	@for os in $(CX_OSES); do \
-		echo "building $$os"; \
-		for arch in $(CX_ARCHS); do \
-			GOOS=$$os GOARCH=$$arch CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o ./bin/$*-$$os-$$arch ./$*/cmd/$*; \
-		done; \
-		if [ $$os = 'windows' ]; then \
-			mv ./bin/$*-$$os-$$arch ./bin/$*-$$os-$$arch.exe; \
-		fi; \
-	done
-
-.PHONY: build-release
-build-release: brig-cross-compile
+
+amd64_GOARCH = amd64
+arm64_GOARCH = arm64
+arm32_GOARCH = arm
+
+amd64_QEMU_IMPORT =
+arm64_QEMU_IMPORT = ADD https://github.com/multiarch/qemu-user-static/releases/download/v3.1.0-2/x86_64_qemu-aarch64-static.tar.gz /usr/bin
+arm32_QEMU_IMPORT = ADD https://github.com/multiarch/qemu-user-static/releases/download/v3.1.0-2/x86_64_qemu-arm-static.tar.gz /usr/bin
+
+amd64_DOCKER_ARCH = amd64
+arm64_DOCKER_ARCH = arm64v8
+arm32_DOCKER_ARCH = arm32v6
+
+amd64_DOCKER_ANNOTATIONS = --os linux --arch amd64
+arm64_DOCKER_ANNOTATIONS = --os linux --arch arm64 --variant v8
+arm32_DOCKER_ANNOTATIONS = --os linux --arch arm
+
+.PHONY: build build-images push-images dockerfiles
+
+define GO_CMD_TARGETS
+# Build command
+$1/rootfs/$2/$1: vendor $(shell find $1 -type f -name '*.go')
+	GOOS=linux GOARCH=$$($2_GOARCH) CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o ./$1/rootfs/$2/$1 ./$1/cmd/$1
+
+build-$2 $1-$2-image: $1/rootfs/$2/$1
+endef
+
+define DOCKER_IMAGE_TARGETS
+# Build image for a single arch
+.PHONY: $1-images push-$1-images
+$1-$2-image: $1/Dockerfile.$2
+	docker build $(DOCKER_BUILD_FLAGS) -t $(DOCKER_REGISTRY)/$1:$2-$(IMAGE_TAG) -f $1/Dockerfile.$2 $1
+
+# Push image for a single arch
+push-$1-$2-image: $1-$2-image
+	docker push $(DOCKER_REGISTRY)/$1:$2-$(IMAGE_TAG)
+
+# Create arch dockerfile
+$1/Dockerfile.$2: $1/Dockerfile
+	@cat $1/Dockerfile | ARCH="$2" QEMU_IMPORT="$$($2_QEMU_IMPORT)" DOCKER_ARCH="$$($2_DOCKER_ARCH)" envsubst '$$$${ARCH} $$$${QEMU_IMPORT} $$$${DOCKER_ARCH}' > $1/Dockerfile.$2
```

Comment:
We'll need to check for the existence of `envsubst` and either install in a platform-agnostic manner if it doesn't exist -- or exit out and inform the user that it is a required dependency to run the targets needing it.

As a Mac user, I didn't have it by default, although I did have the package that contained it (`gettext`).  For me, the key was the latter command from: `brew install gettext && brew link --force gettext`

Reply:
I did the same command :). Alternatively, using something like go templates might be a better idea?

Reply:
Just from half a minute of googeling, `go get github.com/hairyhenderson/gomplate/cmd/gomplate` maybe?

Reply:
Nice -- if that utility works, a go get-able option would be a good choice.